File: Rust/src/challenges/day_07.rs
use super::challenge::{Challenge, Error as ChallengeErr};
use std::collections::{HashMap, HashSet};

const SHINY_GOLD: &str = "shiny gold";

#[derive(Debug)]
struct BagsRules {
    graph: HashMap<String, Vec<(String, i32)>>,
}

impl BagsRules {
    fn reverse(&self) -> BagsRules {
        let mut reversed_graph = BagsRules {
            graph: HashMap::new(),
        };

        for (u, others) in self.graph.iter() {
            for (v, c) in others {
                reversed_graph
                    .graph
                    .entry(v.to_string())
                    .or_insert(vec![])
                    .push((u.to_string(), *c));
            }
        }

        reversed_graph
    }

    fn count_nodes_reachable_from(&self, bag: &String) -> i32 {
        let mut stack = vec![bag];
        let mut visited = HashSet::new();
        visited.insert(bag);
        let mut count = 0;

        while !stack.is_empty() {
            let curr = stack.pop().unwrap();
            count += 1;

            if !self.graph.contains_key(curr) {
                continue;
            }

            for (v, _) in &self.graph[curr] {
                if !visited.contains(&v) {
                    visited.insert(&v);
                    stack.push(&v);
                }
            }
        }

        count
    }

    fn count_bag_colors_containing(&self, bag: &str) -> i32 {
        self.reverse()
            .count_nodes_reachable_from(&String::from(bag))
            - 1
    }

    fn count_bags_inside(&self, bag: &str) -> i32 {
        self.graph[bag]
            .iter()
            .fold(0, |acc, (o, c)| acc + c * (self.count_bags_inside(&o) + 1))
    }

    fn from_rules_text(rules_text: &str) -> BagsRules {
        let mut bags_rules = BagsRules {
            graph: HashMap::new(),
        };

        for bag_rule in rules_text.lines() {
            let (container_bag, bags_inside) = BagsRules::parse_bag_rule(bag_rule);
            let graph_entry = bags_rules.graph.entry(container_bag).or_insert(vec![]);
            for (bag_inside, count) in bags_inside {
                graph_entry.push((bag_inside, count));
            }
        }

        bags_rules
    }

    fn parse_bag_rule(bag_rule: &str) -> (String, Vec<(String, i32)>) {
        let words: Vec<&str> = bag_rule.split_ascii_whitespace().collect();
        let container_bag: String = format!("{} {}", words[0], words[1]);
        let mut bags_inside: Vec<(String, i32)> = vec![];

        if words[4] != "no" {
            for i in (4..words.len()).step_by(4) {
                let count: i32 = words[i].parse().unwrap();
                let other_bag: String = format!("{} {}", words[i + 1], words[i + 2]);
                bags_inside.push((other_bag, count));
            }
        }

        (container_bag, bags_inside)
    }
}

pub struct Solution {}

impl Challenge for Solution {
    fn first_part(&self, bags_rules_text: &str) -> Result<String, ChallengeErr> {
        let bags_rules = BagsRules::from_rules_text(&bags_rules_text);

        Ok(bags_rules
            .count_bag_colors_containing(SHINY_GOLD)
            .to_string())
    }

    fn second_part(&self, bags_rules_text: &str) -> Result<String, ChallengeErr> {
        let bags_rules = BagsRules::from_rules_text(&bags_rules_text);

        Ok(bags_rules.count_bags_inside(SHINY_GOLD).to_string())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn example_count_bag_colors_containing() {
        let test_bags_rules_1 = "light red bags contain 1 bright white bag, 2 muted yellow bags.
dark orange bags contain 3 bright white bags, 4 muted yellow bags.
bright white bags contain 1 shiny gold bag.
muted yellow bags contain 2 shiny gold bags, 9 faded blue bags.
shiny gold bags contain 1 dark olive bag, 2 vibrant plum bags.
dark olive bags contain 3 faded blue bags, 4 dotted black bags.
vibrant plum bags contain 5 faded blue bags, 6 dotted black bags.
faded blue bags contain no other bags.
dotted black bags contain no other bags.";

        let bags_rules = BagsRules::from_rules_text(test_bags_rules_1);

        assert_eq!(bags_rules.count_bag_colors_containing(SHINY_GOLD), 4);
    }

    #[test]
    fn example_count_bags_inside_1() {
        let test_bags_rules_1 = "light red bags contain 1 bright white bag, 2 muted yellow bags.
dark orange bags contain 3 bright white bags, 4 muted yellow bags.
bright white bags contain 1 shiny gold bag.
muted yellow bags contain 2 shiny gold bags, 9 faded blue bags.
shiny gold bags contain 1 dark olive bag, 2 vibrant plum bags.
dark olive bags contain 3 faded blue bags, 4 dotted black bags.
vibrant plum bags contain 5 faded blue bags, 6 dotted black bags.
faded blue bags contain no other bags.
dotted black bags contain no other bags.";

        let bags_rules = BagsRules::from_rules_text(test_bags_rules_1);

        assert_eq!(bags_rules.count_bags_inside(SHINY_GOLD), 32);
    }

    #[test]
    fn example_count_bags_inside_2() {
        let test_bags_rules_2 = "shiny gold bags contain 2 dark red bags.
dark red bags contain 2 dark orange bags.
dark orange bags contain 2 dark yellow bags.
dark yellow bags contain 2 dark green bags.
dark green bags contain 2 dark blue bags.
dark blue bags contain 2 dark violet bags.
dark violet bags contain no other bags.";

        assert_eq!(
            BagsRules::from_rules_text(test_bags_rules_2).count_bags_inside(SHINY_GOLD),
            126
        );
    }
}

File: Python3/aoc2020/solutions/day_07.py
from __future__ import annotations

from collections import defaultdict
from typing import Dict, List, Tuple

from .solution import Solution

BagColor = str


SHINY_GOLD = "shiny gold"


class BagsRules:
    __slots__ = "_graph"

    def __init__(self):
        self._graph: Dict[BagColor,
                          List[Tuple[BagColor, int]]] = defaultdict(list)

    def _add_edge(self, u: BagColor, v: BagColor, c: int) -> None:
        self._graph[u].append((v, c))

    def _reverse(self) -> BagsRules:
        reversed_graph = BagsRules()

        for u, edges in self._graph.items():
            for v, c in edges:
                reversed_graph._add_edge(v, u, c)

        return reversed_graph

    def _count_nodes_reachable_from(self, bag: BagColor) -> int:
        stack = [bag]
        visited = set(stack)
        count = 0

        while stack:
            curr = stack.pop()
            count += 1

            for v, _ in self._graph[curr]:
                if v not in visited:
                    visited.add(v)
                    stack.append(v)

        return count

    def count_bag_colors_containing(self, bag: BagColor) -> int:
        return self._reverse()._count_nodes_reachable_from(bag) - 1

    def count_bags_inside(self, bag: BagColor) -> int:
        total = 0

        for other, count in self._graph[bag]:
            total += count * (self.count_bags_inside(other) + 1)

        return total

    @ classmethod
    def from_rules_text(cls, rules_text: str) -> BagsRules:
        bags_rules = cls()

        for bag_rule_text in rules_text.splitlines():
            container_bag, bags_inside = cls._parse_bag_rule(bag_rule_text)
            for bag_inside, count in bags_inside:
                bags_rules._add_edge(container_bag, bag_inside, count)

        return bags_rules

    @ staticmethod
    def _parse_bag_rule(bag_rule: str) -> Tuple[str, List[Tuple[str, int]]]:
        words = bag_rule.split()
        container_bag = " ".join(words[:2])
        bags_inside: List[Tuple[str, int]] = []

        if words[4] == "no":
            return container_bag, bags_inside

        for i in range(4, len(words), 4):
            count_str, *bag_color = words[i:i+3]
            count = int(count_str, base=10)
            other_bag = " ".join(bag_color)
            bags_inside.append((other_bag, count))

        return container_bag, bags_inside


class Day07(Solution):

    def first_task(self, bags_text: str) -> str:
        bags_rules = BagsRules.from_rules_text(bags_text)

        return str(bags_rules.count_bag_colors_containing(SHINY_GOLD))

    def second_task(self, bags_text: str) -> str:
        bags_rules = BagsRules.from_rules_text(bags_text)

        return str(bags_rules.count_bags_inside(SHINY_GOLD))
